wire-ios-sync-engine/Tests/Source/UserSession/SessionManagerEncryptionAtRestMigrationTests.swift

```diff
@@ -0,0 +1,158 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import LocalAuthentication
+import WireDataModelSupport
+@testable import WireSyncEngine
+
+final class SessionManagerEncryptionAtRestMigrationTests: ZMUserSessionTestsBase {
+
+    var userSessionDelegate: MockUserSessionDelegate!
+    private var activityManager: MockBackgroundActivityManager!
+    private var factory: BackgroundActivityFactory!
+    private var setEncryptionAtRestExpectation: XCTestExpectation?
+
+    private var account: Account {
+        coreDataStack.account
+    }
+
+    private var userSession: ZMUserSession {
+        sut
+    }
+
+    override func setUp() {
+        userSessionDelegate = MockUserSessionDelegate()
+
+        super.setUp()
+
+        activityManager = MockBackgroundActivityManager()
+        factory = BackgroundActivityFactory.shared
+        factory.activityManager = activityManager
+        setEncryptionAtRestExpectation = nil
+    }
+
+    /// This workaround is needed because all tests here are based on assumptions
+    /// that the `managedObjectContext` is changed.
+    /// To remove this workaround, delete this override  and the `mockEARService` should be used instead of
+    /// a real instance of `EARService`.
+    override func createSut() -> ZMUserSession {
+        let earService = EARService(
+            accountID: coreDataStack.account.userIdentifier,
+            databaseContexts: [
+                coreDataStack.viewContext,
+                coreDataStack.syncContext,
+                coreDataStack.searchContext
+            ],
+            canPerformKeyMigration: true,
+            sharedUserDefaults: sharedUserDefaults,
+            authenticationContext: MockAuthenticationContextProtocol()
+        )
+
+        let session = createSut(earService: earService)
+
+        session.delegate = userSessionDelegate
+        userSessionDelegate.prepareForMigrationOnReadyMockMethod = { _, onReady in
+            try onReady(self.uiMOC)
+        }
+
+        return session
+    }
+
+    override func tearDown() {
+        factory = nil
+        activityManager = nil
+
+        try? sut.setEncryptionAtRest(enabled: false, skipMigration: true)
+
+        super.tearDown()
+        userSessionDelegate = nil
+    }
+
+    private func setEncryptionAtRest(enabled: Bool, file: StaticString = #filePath, line: UInt = #line) {
+        try? sut.setEncryptionAtRest(enabled: true, skipMigration: true)
+        XCTAssertTrue(waitForAllGroupsToBeEmpty(withTimeout: 0.5), file: file, line: line)
+    }
+
+    private func login() {
+        syncMOC.performAndWait {
+            simulateLoggedInUser()
+            ModelHelper().createSelfClient(in: syncMOC)
+            syncMOC.saveOrRollback()
+        }
+        try userSession.viewContext.saveOrRollback()
```

Comment:
for unknown reason I had to save explicitly save here for the second test, otherwise the selfUser was missing